tooltips: Group tooltips for smooth transition.

shashank-23002 opened the pull request:
Fixes: #24825

<!-- Describe your pull request here.-->
This PR groups the tooltips for compose box formatting buttons so that the delay only applies to the first button you hover.

This allows two things:
1) Smooth transitions of the tippy between many different reference element targets
2) Elements with tooltips next to each other that have a delay can be "grouped" so they appear to share a timeout, which greatly improves UX.

Previously , the tooltips of formatting buttons were treated independently and each of them had their own delays which doesn't feel smooth while hovering.

Now, the shared timer across the grouped tooltips helps the transition feel smooth and clean.

Fixes: <!-- Issue link, or clear description.-->

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**

Previously: 

https://github.com/zulip/zulip/assets/97145463/acdd50f8-5c0b-4d80-bd47-e097a2eadfd2

Currently: 

https://github.com/zulip/zulip/assets/97145463/a59a503b-8800-4695-bd7b-fc5a27454053



<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

alya commented:
@sayamsamal Are you up for doing a first-round review on this PR? I haven't tested it.

N-Shar-ma commented:
@shashank-23002 Tested locally and have some feedback:
1. The formatting for multiline tooltips (global time and poll buttons) is broken - visible from your screencap too.
2. When disabled on hover, the poll button should have its tooltip appear instantly, which does not happen now. If you read the code deleted in your commit, you'll find the code setting the instant hover delay and comments detailing this as well. Please be careful when deleting code -- take care to not accidentally remove any functionality.

I haven't really reviewed the code, but I'd like you to notice that we use snake_case and not camelCase for naming variables and functions in JavaScript. Make sure to follow these conventions in any code you add too.

shashank-23002 commented:
Thanks @N-Shar-ma for reviewing the code . 
I have implemented the required changes to improve the formatting for the multi-line tooltip.
Also fixed the implementation of the deleted code.
PTAL

shashank-23002 commented:
Hey @sayamsamal @N-Shar-ma , it's been quite a while.
Just a gentle reminder if you could please go for a quick round of review.

sayamsamal commented:
Hi @shashank-23002, my apologies for the delay! Encountered some unexpected issues. I've taken a quick look at your code and will share feedback shortly. Thanks for your patience!

shashank-23002 commented:
Sure @sayamsamal , I will look for other modifications till then. Hoping for a quick review.

sayamsamal commented on web/src/click_handlers.js:
```diff
@@ -388,6 +407,46 @@ export function initialize() {
         $row.find(".markdown_preview").show();
     });
 
+    $("body").on("mouseenter", ".compose_control_button", () => {
+        const buttons = document.querySelectorAll(`
+            .compose_control_button[data-tooltip-template-id],
+            .compose_control_button[data-tippy-content],
+            .compose_control_button_container
+        `);
+
+        const tippyInstances = [...buttons].map((button) => {
+            const tippyContent = button.dataset.tippyContent;
+            const tooltipTemplateId = button.dataset.tooltipTemplateId;
```
There is a major issue with this logic, that being, we are attaching multiple Tippy instances over all the compose control buttons, over and over again when hovering above any one of them. In my manual testing, hovering multiple times would eventually lead to a stage where the tooltips become laggy and unresponsive.

sayamsamal replied:
We should let the Tippyjs library handle the creation and deletion of the tooltips, and thus we should only call the `tippy()` constructor on the appropriate selectors once.

sayamsamal replied:
https://atomiks.github.io/tippyjs/v6/addons/#singleton gives us the example of how we should go on with using the singletons addon.

shashank-23002 replied:
Updated the code so that tippy creation only triggers when the user mouseenters the complete```compose-control-buttons-container``` using tippy() constructor.

sayamsamal commented on web/src/click_handlers.js:
```diff
@@ -388,6 +407,46 @@ export function initialize() {
+            const tippyContent = button.dataset.tippyContent;
+            const tooltipTemplateId = button.dataset.tooltipTemplateId;
+
+            if (tippyContent) {
```
We already have a `get_tooltip_content` method, which does the same, let's use that instead.

sayamsamal commented on web/src/click_handlers.js:
```diff
@@ -388,6 +407,46 @@ export function initialize() {
+            const tippyContent = button.dataset.tippyContent;
+            const tooltipTemplateId = button.dataset.tooltipTemplateId;
+
+            if (tippyContent) {
+                return tippy(button, {
+                    content: tippyContent,
+                });
+            } else if (tooltipTemplateId) {
+                const template = document.querySelector(`#${tooltipTemplateId}`);
+                const templateContent = template ? render_html_template(template.innerHTML) : "";
+
+                return tippy(button, {
+                    content: templateContent,
+                });
```
We already have a `get_tooltip_content` method, which does the same, let's use that instead.

sayamsamal commented on web/src/click_handlers.js:
```diff
@@ -388,6 +407,46 @@ export function initialize() {
+
+        const validTippyInstances = tippyInstances.filter((instance) => instance !== null);
+
+        createSingleton(validTippyInstances, {
+            delay: 500,
+            onTrigger(instance, event) {
+                if (event.currentTarget.classList.contains("disabled-on-hover")) {
+                    instance.setProps({delay: 0});
+                } else {
+                    instance.setProps({delay: 500});
+                }
+            },
```
Also, it's important to know that Singletons are also instances, and thus we are also creating multiple singletons over here, which we want to also avoid. On top of that, we should also take into account the [following](https://atomiks.github.io/tippyjs/v6/addons/#destroy):

> When you call singleton.destroy(), the tippyInstances you passed as an argument will not be destroyed also. They are separate instances that can be reused again elsewhere. You should also destroy the tippy instances upon cleanup.

sayamsamal commented:
@shashank-23002 I have left you some comments where we could definitely use some rethinking the logic. But to summarize, (1) we should avoid using click handlers to add tippy constructors since that leads to duplicate tippy isntances and (2) avoid duplication of code and instead use the pre-exisitng methods whenever possible.

sayamsamal commented:
The most major issue as pointed over [here](https://github.com/zulip/zulip/pull/28387#discussion_r1462424507), still remains unaddressed. Please take another look at `tippyjs.ts` and the [tippyjs docs](https://atomiks.github.io/tippyjs/), to see how we handle `trigger` and `targets` in tippyjs. In short, we shouldn't be using our own event listeners here, and instead pass the `triggers` to tippyjs.

```javascript
const btn_tippys = tippy(".compose_control_button_container, .compose_control_button[data-tooltip-template-id], .compose_control_button[data-tippy-content]");

const singleTon = createSingleton(btn_tippys, {...})
````

sayamsamal commented:
Here is a screencast for you, so that you can better visualize the problem with your approach.

https://github.com/zulip/zulip/assets/82862779/0de4430d-dff7-4900-b61e-4411d092ac2f

shashank-23002 commented:
@sayamsamal it's been quite a while. Just a gentle reminder for you to have a look.

shashank-23002 commented:
@sayamsamal I think I have addressed the feedback and resolved conflicts. I am attaching a screen recording so that you can have a quick review - 


https://github.com/zulip/zulip/assets/97145463/2af554f4-f073-41fa-a2ec-d09308fdfb96

karlstolley commented on web/src/click_handlers.js:
```diff
@@ -1,7 +1,6 @@
 import assert from "minimalistic-assert";
-import tippy from "tippy.js";
-
+import tippy, {createSingleton} from "tippy.js";
```
It's probably better to leave that open line of whitespace above the comment below.

karlstolley commented on web/src/click_handlers.js:
```diff
@@ -894,6 +893,35 @@ export function initialize() {
         }
     });
 
+    // COMPOSE BUTTON TOOLTIPS
```
Can you explain the reasoning behind putting this in the click-handlers file instead of the compose-tooltips file? Or even the Tippy file?

shashank-23002 replied:
Ohh,  initially the logic I used in this PR used click handler for creating the array of tooltip instances of all the buttons. This was enhanced after review from @sayamsamal. 
Although I tried to put this in another file but for other files , the creation of tippy instances is not initiated, hence tooltips are not rendered. 
Probable reason could be that `compose_tooltip.js` file is executed before the tippy library is initiated but I am not sure. Another reason could be the trigger we provide is superimposed by some else trigger which does not create the tooltip array .
Would love to hear your opinion and possible solutions.

karlstolley replied:
Have you experimented with reordering the files so Tippy gets loaded up before `compose_tooltip.js`? That might be a worthy prep commit, given how central Tippy is to the codebase now.

karlstolley commented on web/src/click_handlers.js:
```diff
@@ -894,6 +893,35 @@ export function initialize() {
+    // COMPOSE BUTTON TOOLTIPS
+
+    const compose_buttons = tippy(
+        ".compose_control_button_container, .compose_control_button[data-tooltip-template-id], .compose_control_button[data-tippy-content]",
```
The compose buttons are pretty class-happy, but I wonder if it's worth exploring adding a new shared class for these tooltips? The multiple selectors here look fiddly and brittle to me.

shashank-23002 replied:
I think we should give a try. 
Tried and looks better I guess.

karlstolley commented on web/src/click_handlers.js:
```diff
@@ -894,6 +893,35 @@ export function initialize() {
+            }
+
+            if (event.currentTarget.classList.contains("disabled-on-hover")) {
+                instance.setProps({delay: 0});
```
Is there a reason to not use `INSTANT_HOVER_DELAY` from `tippyjs.ts` here? Or `INTERACTIVE_HOVER_DELAY` inside the else block? We should avoid these kinds of one-off scalar values, whenever possible. Over time, they can make the UI feel inconsistent.

shashank-23002 replied:
Yes there are various dividers between the compose control buttons which re-triggers the `INSTANT_HOVER_DELAY` whenever we try to smoothly transition from one button to another. 
Due to which the `createSingleton()` function is re-triggered everytime a divider is encountered.

shashank-23002 commented:
@karlstolley I have addressed the feedbacks with one question.
PTAL
